File: .github/skills/api-endpoint-generator/endpoint-template.ts
/**
 * endpoint-template.ts
 *
 * Reference implementation of a complete Express.js REST endpoint.
 * Demonstrates the exact pattern taught by the api-endpoint-generator skill:
 *
 *   - Zod schema for input validation and TypeScript type inference
 *   - Typed Request / Response generics (no `any`)
 *   - RFC 7807 Problem Details error format
 *   - Async/await error handling with next(error) delegation
 *   - Immutable response construction
 *   - JSDoc on every public function and type
 *
 * To use as a starting point:
 *   1. Replace "Product" with your resource name (find+replace)
 *   2. Update the Zod schema fields to match your resource
 *   3. Replace `productService.*` calls with your actual service/repository
 *   4. Adjust the route path in the router registration at the bottom
 */

import { Router, Request, Response, NextFunction } from 'express';
import { z } from 'zod';

// =============================================================================
// 1. INPUT SCHEMA — single source of truth for validation AND TypeScript types
// =============================================================================

/**
 * Zod schema for the Create Product request body.
 * Every constraint is expressed here — the handler stays clean.
 */
const CreateProductSchema = z.object({
  name: z
    .string()
    .min(2, 'Name must be at least 2 characters')
    .max(200, 'Name must not exceed 200 characters')
    .trim(),
  description: z
    .string()
    .max(2000, 'Description must not exceed 2000 characters')
    .optional(),
  priceInCents: z
    .number()
    .int('Price must be an integer (cents)')
    .min(1, 'Price must be at least 1 cent')
    .max(999_999_99, 'Price must not exceed $9,999,999.99'),
  category: z.enum(['electronics', 'clothing', 'books', 'food', 'other'], {
    errorMap: () => ({ message: 'Category must be one of: electronics, clothing, books, food, other' }),
  }),
  stockQuantity: z
    .number()
    .int('Stock must be a whole number')
    .min(0, 'Stock cannot be negative')
    .default(0),
  tags: z.array(z.string().max(50)).max(10, 'Cannot have more than 10 tags').default([]),
});

/**
 * Zod schema for the Update Product request body.
 * `.partial()` makes all fields optional — standard for PATCH semantics.
 */
const UpdateProductSchema = CreateProductSchema.partial();

// =============================================================================
// 2. TYPESCRIPT TYPES — inferred from schemas, never hand-written separately
// =============================================================================

/** Input for creating a new product. Inferred from {@link CreateProductSchema}. */
type CreateProductInput = z.infer<typeof CreateProductSchema>;

/** Input for partially updating a product. Inferred from {@link UpdateProductSchema}. */
type UpdateProductInput = z.infer<typeof UpdateProductSchema>;

/** The response shape returned by product endpoints. */
interface ProductResponse {
  readonly id: string;
  readonly name: string;
  readonly description: string | null;
  readonly priceInCents: number;
  readonly category: CreateProductInput['category'];
  readonly stockQuantity: number;
  readonly tags: readonly string[];
  readonly createdAt: string; // ISO 8601
  readonly updatedAt: string; // ISO 8601
}

// =============================================================================
// 3. ROUTE HANDLERS
// =============================================================================

/**
 * POST /products
 *
 * Creates a new product listing.
 *
 * @param req.body - {@link CreateProductInput} — validated by Zod
 * @returns 201 with {@link ProductResponse} of the newly created product
 * @throws 400 if the request body fails schema validation
 * @throws 409 if a product with the same name already exists in this category
 */
async function createProduct(
  req: Request,
  res: Response<ProductResponse>,
  next: NextFunction,
): Promise<void> {
  const parseResult = CreateProductSchema.safeParse(req.body);

  if (!parseResult.success) {
    // Return RFC 7807 Problem Details — never expose raw Zod error objects
    res.status(400).json({
      type: 'https://errors.myapp.com/validation-error',
      title: 'Validation Error',
      status: 400,
      detail: 'The request body contains one or more invalid fields.',
      errors: parseResult.error.flatten().fieldErrors,
    } as unknown as ProductResponse);
    return;
  }

  const input: CreateProductInput = parseResult.data;

  try {
    const product = await productService.create(input);

    // Construct response immutably — never mutate the returned entity
    const response: ProductResponse = {
      id: product.id,
      name: product.name,
      description: product.description ?? null,
      priceInCents: product.priceInCents,
      category: product.category,
      stockQuantity: product.stockQuantity,
      tags: [...product.tags], // defensive copy — return value is read-only
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };

    res.status(201).json(response);
  } catch (error) {
    // Delegate to centralized error-handling middleware
    // Do NOT swallow the error or return 500 manually here
    next(error);
  }
}

/**
 * GET /products/:id
 *
 * Retrieves a single product by its unique identifier.
 *
 * @param req.params.id - The product's UUID
 * @returns 200 with {@link ProductResponse}
 * @throws 400 if `id` is not a valid UUID format
 * @throws 404 if no product with the given id exists
 */
async function getProductById(
  req: Request<{ id: string }>,
  res: Response<ProductResponse>,
  next: NextFunction,
): Promise<void> {
  const idSchema = z.string().uuid('Product id must be a valid UUID');
  const parseResult = idSchema.safeParse(req.params.id);

  if (!parseResult.success) {
    res.status(400).json({
      type: 'https://errors.myapp.com/invalid-id',
      title: 'Invalid ID Format',
      status: 400,
      detail: 'The provided product id is not a valid UUID.',
    } as unknown as ProductResponse);
    return;
  }

  try {
    const product = await productService.findById(parseResult.data);

    if (!product) {
      res.status(404).json({
        type: 'https://errors.myapp.com/not-found',
        title: 'Product Not Found',
        status: 404,
        detail: `No product was found with id '${req.params.id}'.`,
      } as unknown as ProductResponse);
      return;
    }

    const response: ProductResponse = {
      id: product.id,
      name: product.name,
      description: product.description ?? null,
      priceInCents: product.priceInCents,
      category: product.category,
      stockQuantity: product.stockQuantity,
      tags: [...product.tags],
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };

    res.status(200).json(response);
  } catch (error) {
    next(error);
  }
}

/**
 * PATCH /products/:id
 *
 * Partially updates an existing product.
 * Only the fields included in the request body are changed.
 *
 * @param req.params.id - The product's UUID
 * @param req.body - {@link UpdateProductInput} — all fields optional
 * @returns 200 with the updated {@link ProductResponse}
 * @throws 400 if `id` is invalid or body fails validation
 * @throws 404 if no product with the given id exists
 */
async function updateProduct(
  req: Request<{ id: string }>,
  res: Response<ProductResponse>,
  next: NextFunction,
): Promise<void> {
  const idSchema = z.string().uuid();
  const idResult = idSchema.safeParse(req.params.id);

  if (!idResult.success) {
    res.status(400).json({
      type: 'https://errors.myapp.com/invalid-id',
      title: 'Invalid ID Format',
      status: 400,
      detail: 'The provided product id is not a valid UUID.',
    } as unknown as ProductResponse);
    return;
  }

  const bodyResult = UpdateProductSchema.safeParse(req.body);

  if (!bodyResult.success) {
    res.status(400).json({
      type: 'https://errors.myapp.com/validation-error',
      title: 'Validation Error',
      status: 400,
      detail: 'One or more fields in the request body are invalid.',
      errors: bodyResult.error.flatten().fieldErrors,
    } as unknown as ProductResponse);
    return;
  }

  const input: UpdateProductInput = bodyResult.data;

  try {
    const updated = await productService.update(idResult.data, input);

    if (!updated) {
      res.status(404).json({
        type: 'https://errors.myapp.com/not-found',
        title: 'Product Not Found',
        status: 404,
        detail: `No product was found with id '${req.params.id}'.`,
      } as unknown as ProductResponse);
      return;
    }

    const response: ProductResponse = {
      id: updated.id,
      name: updated.name,
      description: updated.description ?? null,
      priceInCents: updated.priceInCents,
      category: updated.category,
      stockQuantity: updated.stockQuantity,
      tags: [...updated.tags],
      createdAt: updated.createdAt.toISOString(),
      updatedAt: updated.updatedAt.toISOString(),
    };

    res.status(200).json(response);
  } catch (error) {
    next(error);
  }
}

// =============================================================================
// 4. ROUTER REGISTRATION — group all product routes under /products
// =============================================================================

export const productRouter = Router();

productRouter.post('/', createProduct);
productRouter.get('/:id', getProductById);
productRouter.patch('/:id', updateProduct);

// =============================================================================
// 5. UNIT TEST STUB — paste into products.test.ts and fill in the assertions
// =============================================================================

/*
import { describe, it, expect, vi, beforeEach } from 'vitest';
import request from 'supertest';
import express from 'express';
import { productRouter } from './endpoint-template';

// Mock the service layer so tests are isolated from the database
vi.mock('./product-service', () => ({
  productService: {
    create: vi.fn(),
    findById: vi.fn(),
    update: vi.fn(),
  },
}));

const app = express();
app.use(express.json());
app.use('/products', productRouter);

describe('POST /products', () => {
  const validBody = {
    name: 'Wireless Headphones',
    priceInCents: 9999,
    category: 'electronics',
  };

  it('returns 201 with the created product for valid input', async () => {
    productService.create.mockResolvedValueOnce({ id: 'abc-123', ...validBody, ... });
    const res = await request(app).post('/products').send(validBody);
    expect(res.status).toBe(201);
    expect(res.body.id).toBeDefined();
  });

  it('returns 400 when required field "name" is missing', async () => {
    const { name, ...bodyWithoutName } = validBody;
    const res = await request(app).post('/products').send(bodyWithoutName);
    expect(res.status).toBe(400);
    expect(res.body.errors.name).toBeDefined();
  });

  it('returns 400 when priceInCents is not an integer', async () => {
    const res = await request(app).post('/products').send({ ...validBody, priceInCents: 9.99 });
    expect(res.status).toBe(400);
  });
});
*/

File: .github/skills/api-endpoint-generator/endpoint-template.py
"""
endpoint-template.py

Reference implementation of a complete FastAPI REST endpoint.
Demonstrates the exact pattern taught by the api-endpoint-generator skill:

  - Pydantic v2 models for request and response validation
  - Field constraints declared in the schema, not the handler
  - HTTPException for expected business errors (4xx)
  - Fully typed function signatures — no plain `dict` returns
  - RFC 7807-style error details via HTTPException.detail
  - Docstrings on every model and endpoint
  - Pytest stub at the bottom using httpx's async client

Usage:
  pip install fastapi pydantic[email] httpx uvicorn
  uvicorn endpoint_template:router --reload

  Or mount the router in your main FastAPI app:
    from endpoint_template import router as product_router
    app.include_router(product_router)
"""

from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from typing import Annotated
from uuid import UUID, uuid4

from fastapi import APIRouter, HTTPException, Path, Query, status
from pydantic import BaseModel, EmailStr, Field, field_validator, model_validator

# =============================================================================
# 1. ENUMERATIONS — define valid discrete values as enums, not raw strings
# =============================================================================

class ProductCategory(StrEnum):
    """Valid categories for a product listing."""
    ELECTRONICS = "electronics"
    CLOTHING = "clothing"
    BOOKS = "books"
    FOOD = "food"
    OTHER = "other"

# =============================================================================
# 2. REQUEST SCHEMAS — Pydantic v2 models with full field constraints
# =============================================================================

class CreateProductRequest(BaseModel):
    """
    Request body for creating a new product listing.

    All validation constraints are declared at the field level.
    FastAPI returns HTTP 422 automatically if validation fails.
    """

    name: Annotated[str, Field(min_length=2, max_length=200, examples=["Wireless Headphones"])]
    description: Annotated[str | None, Field(max_length=2000, default=None)]
    price_in_cents: Annotated[int, Field(ge=1, le=99_999_999, description="Price in cents (e.g. 999 = $9.99)")]
    category: ProductCategory
    stock_quantity: Annotated[int, Field(ge=0, default=0)]
    tags: Annotated[list[str], Field(max_length=10, default_factory=list)]

    @field_validator("name")
    @classmethod
    def name_must_not_be_blank(cls, value: str) -> str:
        """Reject names that are only whitespace after stripping."""
        stripped = value.strip()
        if not stripped:
            raise ValueError("Name must not be blank or whitespace only")
        return stripped

    @field_validator("tags")
    @classmethod
    def tags_must_be_non_empty_strings(cls, value: list[str]) -> list[str]:
        """Each tag must be a non-empty string after stripping."""
        validated = [t.strip() for t in value]
        if any(not t for t in validated):
            raise ValueError("Each tag must be a non-empty string")
        return validated


class UpdateProductRequest(BaseModel):
    """
    Request body for partially updating an existing product.

    All fields are optional — only the provided fields are changed.
    This implements PATCH semantics correctly.
    """

    name: Annotated[str | None, Field(min_length=2, max_length=200, default=None)]
    description: str | None = None
    price_in_cents: Annotated[int | None, Field(ge=1, le=99_999_999, default=None)]
    category: ProductCategory | None = None
    stock_quantity: Annotated[int | None, Field(ge=0, default=None)]
    tags: list[str] | None = None

    @model_validator(mode="after")
    def at_least_one_field_required(self) -> "UpdateProductRequest":
        """Ensure the PATCH body is not empty."""
        if all(v is None for v in self.model_dump().values()):
            raise ValueError("At least one field must be provided for update")
        return self


# =============================================================================
# 3. RESPONSE SCHEMAS — always return a typed model, never a raw dict
# =============================================================================

class ProductResponse(BaseModel):
    """The product representation returned by the API."""

    id: UUID
    name: str
    description: str | None
    price_in_cents: int
    category: ProductCategory
    stock_quantity: int
    tags: list[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}  # Allows construction from ORM objects


class ProductListResponse(BaseModel):
    """Paginated list of products."""

    items: list[ProductResponse]
    total: int
    page: int
    page_size: int


# =============================================================================
# 4. ROUTE HANDLERS — each handler is a single async function
# =============================================================================

router = APIRouter(prefix="/products", tags=["products"])


@router.post(
    "/",
    response_model=ProductResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new product listing",
    responses={
        409: {"description": "A product with this name already exists in this category"},
        422: {"description": "Request body validation failed"},
    },
)
async def create_product(payload: CreateProductRequest) -> ProductResponse:
    """
    Create a new product listing.

    - **name**: 2–200 characters, whitespace stripped
    - **price_in_cents**: Must be a positive integer (e.g., 999 = $9.99)
    - **category**: Must be one of the predefined categories
    - **tags**: Up to 10 non-empty tags
    """
    try:
        # Replace with your actual service/repository call
        created = await product_service.create(
            name=payload.name,
            description=payload.description,
            price_in_cents=payload.price_in_cents,
            category=payload.category,
            stock_quantity=payload.stock_quantity,
            tags=payload.tags,
        )
    except DuplicateProductError as exc:
        # Expected business error — raise as HTTPException, not a generic Exception
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "type": "https://errors.myapp.com/duplicate-product",
                "title": "Duplicate Product",
                "status": 409,
                "detail": f"A product named '{payload.name}' already exists in category '{payload.category}'.",
            },
        ) from exc

    # Return a Pydantic model — FastAPI serializes it automatically
    # from_orm=True (model_config above) lets this work with SQLAlchemy models too
    return ProductResponse.model_validate(created)


@router.get(
    "/{product_id}",
    response_model=ProductResponse,
    summary="Retrieve a product by ID",
    responses={
        404: {"description": "Product not found"},
    },
)
async def get_product(
    product_id: Annotated[UUID, Path(description="The product's unique identifier")],
) -> ProductResponse:
    """
    Retrieve a single product by its UUID.

    FastAPI automatically validates that `product_id` is a valid UUID format
    and returns 422 if it is not — no manual validation needed here.
    """
    product = await product_service.find_by_id(product_id)

    if product is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={
                "type": "https://errors.myapp.com/not-found",
                "title": "Product Not Found",
                "status": 404,
                "detail": f"No product was found with id '{product_id}'.",
            },
        )

    return ProductResponse.model_validate(product)


@router.get(
    "/",
    response_model=ProductListResponse,
    summary="List products with optional filtering",
)
async def list_products(
    category: Annotated[ProductCategory | None, Query(description="Filter by category")] = None,
    page: Annotated[int, Query(ge=1, description="Page number (1-based)")] = 1,
    page_size: Annotated[int, Query(ge=1, le=100, description="Results per page")] = 20,
) -> ProductListResponse:
    """
    Return a paginated list of products, optionally filtered by category.

    Query parameters:
    - **category**: Optional. Filter to products in this category.
    - **page**: Page number, starting at 1.
    - **page_size**: Number of results per page (max 100).
    """
    items, total = await product_service.list_all(
        category=category,
        offset=(page - 1) * page_size,
        limit=page_size,
    )

    return ProductListResponse(
        items=[ProductResponse.model_validate(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.patch(
    "/{product_id}",
    response_model=ProductResponse,
    summary="Partially update an existing product",
    responses={
        404: {"description": "Product not found"},
    },
)
async def update_product(
    product_id: Annotated[UUID, Path(description="The product's unique identifier")],
    payload: UpdateProductRequest,
) -> ProductResponse:
    """
    Partially update a product. Only fields included in the request body are changed.

    Send an empty body `{}` to receive a 422 — at least one field is required.
    """
    updated = await product_service.update(
        product_id=product_id,
        # Use model_dump(exclude_none=True) to apply only the provided fields
        updates=payload.model_dump(exclude_none=True),
    )

    if updated is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={
                "type": "https://errors.myapp.com/not-found",
                "title": "Product Not Found",
                "status": 404,
                "detail": f"No product was found with id '{product_id}'.",
            },
        )

    return ProductResponse.model_validate(updated)


# =============================================================================
# 5. PYTEST STUB — paste into test_products.py and complete the assertions
# =============================================================================

"""
# test_products.py

import pytest
from httpx import AsyncClient, ASGITransport
from fastapi import FastAPI
from unittest.mock import AsyncMock, patch
from uuid import uuid4
from datetime import datetime, timezone

from endpoint_template import router

app = FastAPI()
app.include_router(router)

SAMPLE_PRODUCT_ID = uuid4()
SAMPLE_PRODUCT = {
    "id": str(SAMPLE_PRODUCT_ID),
    "name": "Wireless Headphones",
    "description": None,
    "price_in_cents": 9999,
    "category": "electronics",
    "stock_quantity": 100,
    "tags": [],
    "created_at": datetime.now(timezone.utc).isoformat(),
    "updated_at": datetime.now(timezone.utc).isoformat(),
}

@pytest.fixture
async def client():
    async with AsyncClient(transport=ASGITransport(app=app), base_url="http://test") as ac:
        yield ac

@pytest.mark.anyio
async def test_create_product_returns_201(client):
    with patch("endpoint_template.product_service.create", new_callable=AsyncMock) as mock_create:
        mock_create.return_value = SAMPLE_PRODUCT
        response = await client.post("/products/", json={
            "name": "Wireless Headphones",
            "price_in_cents": 9999,
            "category": "electronics",
        })
    assert response.status_code == 201
    assert response.json()["name"] == "Wireless Headphones"

@pytest.mark.anyio
async def test_create_product_returns_422_for_missing_name(client):
    response = await client.post("/products/", json={
        "price_in_cents": 9999,
        "category": "electronics",
    })
    assert response.status_code == 422

@pytest.mark.anyio
async def test_create_product_returns_422_for_negative_price(client):
    response = await client.post("/products/", json={
        "name": "Widget",
        "price_in_cents": -1,
        "category": "other",
    })
    assert response.status_code == 422

@pytest.mark.anyio
async def test_get_product_returns_404_for_unknown_id(client):
    with patch("endpoint_template.product_service.find_by_id", new_callable=AsyncMock) as mock_find:
        mock_find.return_value = None
        response = await client.get(f"/products/{uuid4()}")
    assert response.status_code == 404
"""
